### PR TITLE
pb: Do not exceed the specified total transfer bytes

### LINKED ISSUE #2684: Progress bar is not updated properly when uploading a multipart file
## Expected behavior
Progress bar correctly reflects the update status

## Actual behavior
```
15:14 $ mc cp /mnt/p10/vadmeste/backup/512mb myminio/testbucket/
.../backup/512mb:  226.46 MB / 512.00 MB ┃▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓█░░░░░░░░░░░░░░░░░░░░░┃  44.23% 46.78 MB/s 4s
15:14 $

```

## Steps to reproduce the behavior
1. Upload a file of 512 megabytes

## mc version
master

## System information
All

### PR DESCRIPTION
**Pull request by vadmeste**
mc, via minio-go, can retries some requests for many reasons, such as, network
disconnect. This commit ensures that the progress bar does not show transfered
bytes greater then the total transfered bytes to avoid confusing users.

Fixes #2684

### CODE REVIEW

File: cmd/progress-bar.go [hunk 1]
```diff
@@ -102,18 +96,17 @@ func (p *progressBar) Set64(length int64) *progressBar {
...
-		return p.ProgressBar.Total - offset, nil
-	}
-	return 0, errors.New("invalid argument")
+func (p *progressBar) Read(buf []byte) (n int, err error) {
+	defer func() {
+		// After updating the internal progress bar, make sure that its
+		// current progress doesn't exceed the specified total progress
+		currentProgress := p.ProgressBar.Get()
+		if currentProgress > p.ProgressBar.Total {
+			p.ProgressBar.Set64(p.ProgressBar.Total)
```
Comment (ebozduman) [inline cmd/progress-bar.go]: I think this logic has a problem: After so many retries, we'll reach the total but we are not going to be completely done yet. That means, end-user will see the same total number for sometime while the copy is still continuing and the displayed number will show the total size until the copy is  really done...

**Comment by harshavardhana** [comment]
Seek would still be needed if you want to Seek back the stream @vadmeste so can you add it back ?

**Comment by vadmeste** [comment]
> Seek would still be needed if you want to Seek back the stream @vadmeste so can you add it back ?

@harshavardhana I think the logic of Seek is wrong here, because parallel uploads (like multipart) have access to the same progress bar. Calling Seek() will make the progress bar confused. A simple case is Seek(0, 0) which makes the progress bar looks like it is starting from the beginning although other uploads could be already successfully completed.

I think a cleaner way to fix this is to have each upload has its own progress reporting and it connects itself to the general progress bar (by the way, this can be a first step for showing multiple line progress)

But this PR is urgent before the release. mc master currently wrongly reports progress of a multipart upload.

**Comment by harshavardhana** [comment]
> @harshavardhana I think the logic of Seek is wrong here, because parallel uploads (like multipart) have access to the same progress bar. Calling Seek() by one upload will make the progress bar confused. A simple case is Seek(0, 0) which makes the progress bar looks like it is starting from the beginning although other uploads could be already successfully completed.

@vadmeste okay understood

File: cmd/progress-bar.go (hunk 1 above)
Comment (vadmeste) [reply cmd/progress-bar.go]: > I think this logic has a problem: After so many retries, we'll reach the total but we are not going to be completely done yet. That means, the progress bar will be still moving forward, while the displayed number is stuck at the total size for a while until the copy is really done.

mc will still show an error when the upload fails even after having the progress bar arrived to 100%. 

This actually can happen, for example, during complete multipart upload.

In the future, we can make the progress bar reduces its progress when a part upload fails

**Comment by harshavardhana** [comment]
In essence we can never go back because multiparts have a single progress reader and they are all parallel..

**Comment by vadmeste** [comment]
> In essence we can never go back because multiparts have a single progress reader and they are all parallel..

Yes, but in the future, we can separate upload progress from the UI progress bar, so the UI progress bar counts the progress of all registered uploaders, this looks like doable

**Pull request merged**